Cash Flow and Fix It Strategies

mgreen-simplethread opened the pull request:
This adds the last major missing piece of functionality to the application - analysis of the user's transactions and the suggestion of strategies for improving their cash flow based on predefined criteria.

There are 2 different types of recommendations: Strategies, and Fix It Strategies. They follow different sets of rules:

## Strategies

Strategies are displayed based on the existence of expenses in categories to which the strategies pertain. If the user has even one expense in a category with a strategy, that strategy is displayed. They have brief suggestions and most have off-site informational links that open in a new tab.

+ The store analyzes the entire dataset that the user has entered when considering which strategies to recommend.
+ The store reduces the set of transactions into a unique (deduplicated) list of transaction categories.
+ Those transaction categories are matched against categories that have a `strategy` property. Any and all matching strategies are returned and displayed in the UI.

A couple strategies centered around savings and healthcare expenses will show up in the opposite case - e.g. if the user has no record of putting money into emergency savings, they'll get a strategy recommending that they start putting some money away each time they get paid.

![cfpb-strategies](https://user-images.githubusercontent.com/45372816/78833232-84f8d780-79ba-11ea-9cbf-697d5177b24d.gif)


## Fix It Strategies

Fix-it strategies are actionable within the app. They have buttons that take the user to the expense that triggered them, allowing them to edit that expense according to the strategy's advice.

+ Fix-it strategies are only accessible to the user if they end the week with a negative balance. A call to action button will display in the calendar's details section above the list of transactions if this happens, allowing them to view the strategies.
+ Only the transactions in the currently selected week are analyzed at a time
+ A more complex set of criteria is used to select a max of 3 Fix It strategies to display to the user:
  + If the user has one or more housing expenses (rent or mortgage payments) in the selected week, the largest one is selected, and a corresponding strategy is returned
  + The user's largest regularly billed expense of that week is found and matched against a list of categories. If it matches, the corresponding strategy is returned.
  + The user's largest ad-hoc expense (an expense that doesn't have a fixed, recurring bill associated with it) is found and matched against a list of categories. If it matches one, the corresponding strategy is returned.

There can be anywhere from 0 to 3 Fix It strategies displayed.

Demo of moving a student loan payment to earlier in the month. Balance goes from -$100 to $0 at the end of the week. Not great, but better.

![cfpb-fix-it-strategies](https://user-images.githubusercontent.com/45372816/78833286-9cd05b80-79ba-11ea-9be2-b98044559082.gif)


## Staging

This is available up on the Heroku staging server: https://cfgov-refresh-staging.herokuapp.com/mmt-my-money-calendar/calendar

I'll push this to a CFPB dev server once I merge. Things are slow there right now, and I still have a server all to myself for the main `my-money-calendar` branch.

meissadia commented:
@mgreen-simplethread This is looking fantastic, really cool to see it all come together!

A few observations.

## Strategies
- Explore CFPB's Military Financial Resources
  - servicemembers => service members
- Explore CareerOneStop Job Training Opportunities
  - no description? 
- Explore Level Payment Plans for Utilities
  - no link to more details?
  - multiple utility entries => multiple Level Payment strategies
    - ex. 1 Natural Gas + 1 Water  + 1 Electricity expenses = 3 instances of the same strategy
- Explore Low Cost Phone and Internet Services
  - no link
     - https://www.fcc.gov/consumers/guides/lifeline-support-affordable-communications
  - multiple phone bill expenses generate duplicate strategy instances
- Compare the Rates of Other Insurance Companies
  - Most car Insurance providers offer _quote_ => **quotes**

## Fix-it Strategies
  - "Below are the transactions you made." 
     - Only shows transactions for which actionable strategies have been assigned?
     - "Below are transactions that may be adjusted to improve cash flow"?
       - Maybe the original language is fine but I felt a disconnect when I only saw one actionable transaction listed. 
  - After editing a Fix-It transaction and returning to the Fix-It page, the old category was displayed
    - Edited "Fun Money" transaction, which was the largest non-obligatory spending
    - Returned to Fix-It page
    - Button showed "Edit Eating Out" but "Fun Money" was still shown in the description
    - Everything looked correct after a page reload
    - I didn't click the button, so I'm not sure if it would have taken me to the 'Fun Money' transaction or the 'Eating Out' transaction. 
     ![Fun-Money-Sticks](https://user-images.githubusercontent.com/2592907/78940775-a5389d00-7a73-11ea-90ff-6065c2321180.gif)

mgreen-simplethread commented:
@meissadia I fixed just about everything:

__Strategies:__

+ Typos in strategy copy are fixed where applicable
  + Some things are intentional, like the missing description in Job Training Opportunities. That didn't have any copy associated with it in the source document I was given. Ditto for no link in level payment plans for utilities.
+ Fixed duplicate strategies for different utility types by giving all the strategies a unique ID slug and de-duplicating based on it before displaying.
+ Added missing link for lifeline communication plans

__Fix It Strategies:__

+ I'll speak with Ryan about the copy stuff you highlighted. Revisions there will need to come from him.
+ Fixed the issue with the incorrect category name staying in the description after fixing an expense. A few of them use a template to insert the category name into the message. I was overwriting the template function with its resulting string after the first time it was executed, so there was no template function to call anymore after the first time.

Fixes have been deployed to stage. Thanks for your feedback! You've been a huge help with QAing this app.
